### ISSUE TITLE
Export/import [$30]

### ISSUE THREAD

**Issue by jospoortvliet** [nextcloud/deck#87]
So Desk is awesome and we have decided to start using it for our marketing coordination (hi @NinaNC !)

One thing I'd like to see, but probably crazy, is a way to export boards. Both because I'm always worried something breaks but also it'd be nice to have them available off line.

Even nicer would be if the actual boards could be stored in markdown-ish text files on Nextcloud itself so I have full access to them and can keep working when flying but I realize that that is probably a bit too far fetched :roll_eyes:

**Comment by dan-nkl** [comment]
Would love to see a Export/Import function!

**Comment by wgregori** [comment]
speaking of import function. It would be nice to be able to import a templated structure of tasks to be completed.  In my business I have a series of tasks that must be tracked for every transaction.  The tasks are the same for each transaction.  It would be great to be able to import the template and just go.  Thanks!

**Comment by xenithorb** [comment]
I'm looking for this feature mostly because I'm concerned about putting a lot of effort into setting up and detailing many boards for myself, and then somehow losing all of it through an app-update or some nextcloud quirk. I'd like to not only be able to export it in an importable format, but in some kind of flat-file way too - perhaps a spreadsheet - so that I know my work and time I put into this is safe.

**Comment by dairaine** [comment]
+1 on the possibiliy to export decks, e.g. in a simple csv.

**Comment by juliushaertl** [comment]
@dairaine Please use the [Github Reactions](https://github.com/blog/2119-add-reactions-to-pull-requests-issues-and-comments) feature instead of commenting that also need a specific feature. That way we can see your support without notifying all users subscribed to a given issue.

Besides that adding a bounty at https://www.bountysource.com/issues/43773651-request-export-import might encourage community members to jump on and implement a specific feature. :wink: 

A WIP implementation for user data export is in #470

**Comment by Nebri** [comment]
@juliushaertl How should this be implemented? My initial thoughts is to create a page in the admin to export various boards / everything (for example a list of check boxes that represents the boards, the stacks, and cards in a tree) or to have an export button on the board pages themselves available to admins.

And yes I'm loving the "bounty hunter" role that I have found via Bounty source. It makes priorities / features that are in demand very clear for a contributor like myself.

**Comment by juliushaertl** [comment]
I think we will limit import/export to boards, so we can add the export action to the 3-dots menu in the board list. The should always create a new board then. We could put that one at the bottom of the app navigation (left sidebar).

**Comment by karloluiten** [comment]
As long as there is no export/import, is there a way to manually transfer the data to a new server? Via sql for example? I changed from VPS hosted NC to a server at home and would like to transfer this data as well.

**Comment by woosting** [comment]
Thanks for making exporting possible (for those who do not know this yet; [Exports are possible from the command line using the corresponding OCC command](https://help.nextcloud.com/t/how-to-export-user-data-to-json-by-occ/35664/2?u=woosting)!

Also implementing its counterpart (an import option), would make Deck even better as the combination would provide a Deck-content migration (and even backup) functionality.

@juliushaertl I see you removed this from the Backlog? Does this mean it is off scope for the near future, or the other way around; has it been picked up by someone?

I'm stressing this as not having an import (or better yet: migration) feature, nor well-known procedure, keeps people in my organization (including myself) from using Deck for mission critical planning.

**Comment by blitzdesigner** [comment]
to give this issue maybe a push, the bounty went up to 20$ :-)

**Comment by gary-kim** [comment]
I'm thinking of working on the import function. Would that still be eligible for the bounty?

**Comment by putt1ck** [comment]
We're interested in working on it in relation to Trello exports i.e. a migration tool - but wouldn't compete for the bounty because it's client work :)

**Comment by gary-kim** [comment]
That's nice. I was thinking of adding an `occ` command for now so if we go that route, maybe we should also have that command allow the admin to specify whether the export is from another Nextcloud instance or Trello?
Should not be too difficult to then reuse the code for `occ` imports for the web interface.

**Comment by putt1ck** [comment]
Was assuming occ and the JSON export from Trello (i.e. assuming those migrating are not paying for business class or enterprise). Also assumed board by board export/import for much the same reason - and that issues were more likely to arise with a bulk import.

**Comment by spackmat** [comment]
The export feature is a nice begin, thanks for that!

But implementing it only as a cli command and make it always export all boards of a User is quite limiting. The missing import (or even duplication) of a board is the last thing what prevents me from using Deck.

**Comment by putt1ck** [comment]
Isn't cloning (duplication) dealt with in https://github.com/nextcloud/deck/issues/119 ?

**Comment by spackmat** [comment]
Yes, cloning belongs to #119. I mentioned that, because this would also be solved with a working import/export of single boards available in the backend.

**Comment by daveriedstra** [comment]
Hi! No action on this since end of October, it looks like. Trello is now prompting users to "update their account" & accept new Atlassian terms by April 1. Really wish I could import my Trello data to Deck so I don't have to!

**Comment by putt1ck** [comment]
It's still on our list, just there's a lot of things on it! I think the Deck API is more or less functionally complete from the import perspective, so our main focus would be the translation between Trello output and Deck features

**Comment by juliushaertl** [comment]
Trello import has already been discussed in https://github.com/nextcloud/deck/issues/30 but as mentioned it is not on the road map currently. However if someone is willing to contribute, I'd be happy to integrate that.

From a UI perspective we could have a simple button to upload a json file like calendar does for ics files:
![image](https://user-images.githubusercontent.com/3404133/76620650-efd70000-652d-11ea-9707-dd5401d7f900.png)

I would prefer to do this in several steps, so first we make sure that json exported from deck can be imported properly and in a second step we add some detection for the trello export format and map the values properly to deck parts then.

**Comment by ahvahsky2008** [comment]
> Trello import has already been discussed in #30 but as mentioned it is not on the road map currently. However if someone is willing to contribute, I'd be happy to integrate that.
> 
> From a UI perspective we could have a simple button to upload a json file like calendar does for ics files:
> ![image](https://user-images.githubusercontent.com/3404133/76620650-efd70000-652d-11ea-9707-dd5401d7f900.png)
> 
> I would prefer to do this in several steps, so first we make sure that json exported from deck can be imported properly and in a second step we add some detection for the trello export format and map the values properly to deck parts then.

How download it?

**Comment by ahvahsky2008** [comment]
I want simple convert trello json to ics

**Comment by juliushaertl** [comment]
Nothing like that is implemented yet, this is.the issue where we discuss possible solutions.

**Comment by juliushaertl** [comment]
Hey there,

I'd just like to let you know that due to the [ongoing issue with Bountysource](https://help.nextcloud.com/t/dont-use-bountysource-anymore/84943) we're discussing ways to move to a different platform. As it looks right now, Bountysource might claim some of the money that was contributed by the community. We will try our best to prevent that. Historically we – the Nextcloud GmbH employees – simply put the claimed bounties back to other tickets until one was solved by a community member. So the money was always *from the community* and *for the community*. We'll therefore move to a better platform and currently don't post any new bounties until this situation is resolved. The current goal is to move/donate the money to [Nextcloud include](https://nextcloud.com/include/) but details will follow in the [forum post](https://help.nextcloud.com/t/dont-use-bountysource-anymore/84943).

Hope this makes sense. If you have any questions please ask them on the forum. This comment will be copy-pasted to all open issues with bounties.

**Comment by svbergerem** [comment]
I wrote a small python script that uses the API to copy your boards from one Nextcloud instance to another. It doesn't handle assigned users, permissions or attachments. Nevertheless, I thought it could be useful for some of you. I'm not planning to do any maintenance.

https://gist.github.com/svbergerem/5914d7f87764901aefddba125af99938

**Comment by Rik44444** [comment]
hi, I just want to add that for me this is the only thing keeping me from switching to Deck. I have all my todos in Trello atm, and switching will already be a hassle, so i do not want to risk having to do that again in the future. Happy to add to the bounty, is the Bountysource thing working again (i saw a negative comment about it)?

**Comment by sootstone** [comment]
@Rik44444 Though it is still possible, it is recommended not to add to the bounty at this time as Nextcloud are looking at alternative solutions after BountySource planned to make bounties expire after 2 years (though they did then retract those plans).

Which does make it difficult for us non-contributors to help in any way. I too am in a similar position where having manually copied everything from one instance to another in the past, I will need to migrate to another Nextcloud instance soon and am dreading the thought of having to repeat that process again! Though I will try **svbergerem's** script when it comes to it.

I think Deck is great and improving all the time, and I will still use it over any alternative platforms despite this issue. I'm not really in any position to ask this as I've made no contribution to this project, but it would be great just to have an update on how this is coming along?

**Comment by sootstone** [comment]
Also not that it really matters now but the bounty is currently at $60 as of this time.

**Comment by mstyp** [comment]
Please either fix this or impliment federation.

**Comment by stefan-niedermann** [comment]
@mstyp as i told you in the other issue: Please use **[GitHub reactions](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/)** 👍 to show that you are affected by the same issue. Please don't comment if you have no relevant information to add!

**Comment by ALutz273** [comment]
Now it is already $65.00 USD ;-)

**Comment by rhinozeross** [comment]
Hi Folks!

I'm using Deck for Years now and i really like to use it, its really helping me with my chaos:)
sadly there is still no easy way to Export/Import Boards..

As i already had to switch to different Hosters, sadly i had to manually re-build all of my boards:/ thats really Time-eating and a lot of work:/
is there still no way to export it to a file and to import my boards into a new Nextcloud from (the same) File?

How can i pay some Bounty e.g. to get this necissary Feature? is there a way?

With kind regards
Sebastian

**Comment by varac** [comment]
I already just lost data in one of my shared boards (whole list got accidenitally removed). The lack of a backup/restore (import/export) is quite a bummer, other apps like calendar/contacts support it and that already saved my ass in the past.
Please consider prioritizing it, a lot of ppl voted for this feature already.
I'm happy to top the bounty but it would be good to have a trustworthy bounty provider first (any news on that ?).

**Comment by varac** [comment]
This is the export script that I use: https://0xacab.org/-/snippets/928. It pulls all boards, details and cards and dumps them as json to stdout. Still better than nothing.

**Comment by th23x** [comment]
@varac Thanks for this snippet! Do you by any chance know such a script also for the restore of such dump?

**Comment by varac** [comment]
> @varac Thanks for this snippet! Do you by any chance know such a script also for the restore of such dump?

@th23x The only reference I have is the original scipt (see https://github.com/nextcloud/deck/issues/87#issuecomment-648727278), which I only modified for my purposes. The original script copies over the decks from one to another NC instance, so it shouldn't be hard to modify the script to restore a backup to the original server. But I have not tried that so far.

**Comment by Klemet** [comment]
I just wanted to propose a small workaround for people worried of losing most of their information contained in their boards, like I do.

You can **download your Deck boards as tasks list, and import them as tasks list if needed**.

For this, you need three apps : Calendar, Tasks and Decks.

1. Go on Deck; open the options of Deck (bottom-left corner); enable "Show boards in calendars/tasks"
2. Go in tasks; for each of your deck, in the left menu of Tasks, click on the three dots and choose "download". **This will download your board as a task list file** (ics/Icalendar file).
3. If you lose data concerning your board, go in the Calendar app; open the option of Calendar (bottom-left corner); select "import calendar"; choose the Icalendar file corresponding to your board that you backup up; *et voila*. Now, you can transform those tasks back into board. Comments on each card are kept; I don't know about attached files.

This solution is, of course, very imperfect. I don't know how to automatize it; and if you do import your backup back into your nextcloud after a data loss, you still have to transform the board that is under the form of tasks back into the form of a board.

An import/export function directly for Decks would be wonderful indeed, and I'm really looking forward to it.

**Comment by dtcg** [comment]
> I just wanted to propose a small workaround for people worried of losing most of their information contained in their boards, like I do.
> 
> You can **download your Deck boards as tasks list, and import them as tasks list if needed**.
> 
> For this, you need three apps : Calendar, Tasks and Decks.
> 
>     1. Go on Deck; open the options of Deck (bottom-left corner); enable "Show boards in calendars/tasks"
> 
>     2. Go in tasks; for each of your deck, in the left menu of Tasks, click on the three dots and choose "download". **This will download your board as a task list file** (ics/Icalendar file).
> 
>     3. If you lose data concerning your board, go in the Calendar app; open the option of Calendar (bottom-left corner); select "import calendar"; choose the Icalendar file corresponding to your board that you backup up; _et voila_. Now, you can transform those tasks back into board. Comments on each card are kept; I don't know about attached files.
> 
> 
> This solution is, of course, very imperfect. I don't know how to automatize it; and if you do import your backup back into your nextcloud after a data loss, you still have to transform the board that is under the form of tasks back into the form of a board.
> 
> An import/export function directly for Decks would be wonderful indeed, and I'm really looking forward to it.

Hey @Klemet , it all works apart from the last step. 
So my imported boards all show up as tasks in the tasks app. Now, how to "transform tasks back into boards"? Manually?

**Comment by Klemet** [comment]
Yeah, this has to be done manually, @dtcg. That is why I consider this workaround as very imperfect. Still, you can copy/paste the cards's titles, tags and comments, allowing you to recreate the structure of your boads if needed.

**Comment by vitormattos** [comment]
I created [this](https://github.com/LibreCodeCoop/deck_import_trello) repository. I just rewrote [this](https://github.com/mclang/nextcloud-deck-import-trello-json) project in PHP.

Suggestions for next steps:

- [ ] Make a Deck Command
- [ ] Remove API requests and transform them into requests for Services classes
- [ ] Implement in json configuration a from/to Trello accounts for Deck accounts

Can I continue with the next steps?

**Comment by vitormattos** [comment]
I created a pull request to import a json file exported from a Trello board. I will be happy with the comments and code review.

**Comment by bugzy** [comment]
I have been waiting on this for a while so I decided to search github for implementations and it looks like some other folks have been tackling this issue in various ways. just thought to post this here for inspiration and ideas towards an official solution (Note: I have NOT tested any of these applications):

- https://github.com/newroco/DeckImportExport
- https://github.com/keitalbame/nextcloud-deck-export-import
- https://github.com/filipizydorczyk/next-cloud-deck-backup
- https://github.com/mclang/nextcloud-deck-import-trello-json
- https://github.com/jonasof/kb-to-nd

**Comment by vitormattos** [comment]
@Bugsbane in the pull request I sent I already imported everything that is possible from the Trello to the Deck. What is pending is the code review, who can cooperate will be good.

#3182

**Comment by vitormattos** [comment]
@bugzy in the pull request I sent I already imported everything that is possible from the Trello to the Deck. What is pending is the code review, who can cooperate will be good.

#3182

**Comment by bugzy** [comment]
@vitormattos great to hear that. Am I understanding correctly that your pull request addresses only trello JSON import and not NC Deck JSON import?

**Comment by vitormattos** [comment]
@bugzy I'm happy to help! :-D

The import I implemented accepts for now two methods:

* Trello JSON files. It has a limit of 1000 actions.
* Trello API. This is the most recommendable, it matters everything that is possible from Trello.

There are more details in the PR description and in the documentation that I implemented.

This PR I made opens the possibility to import from other board systems like Asana for example.

**Comment by bugzy** [comment]
@vitormattos  thanks for the explanation.

**Comment by sunjam** [comment]
My honest question: Will Deck be able to import and export from Deck?

I see Trello import support, but wondering about Deck Import and Export... should a different issue be opened?

**Comment by vitormattos** [comment]
Hi @sunjam ,
To import it is necessary to define from which system the data will be imported.
With #3182 I implemented the tool to import from Trello Json and Trello API and to make easy and possible implement the ETL to import probably from any board system. 
The export already was implemented as a command (`occ deck:export`) and the frontend is WIP on #3065.

**Comment by juliushaertl** [comment]
I've wrapped the remaining parts into smaller and separated tickets 

https://github.com/nextcloud/deck/issues?q=is%3Aopen+label%3Aenhancement+label%3A%22feature%3A+export%2Fimport%22

**Comment by Majed610** [comment]
I have one quick question, because my coding-experience is not the best. 

In OCC after using the command deck:export (which use export a JSON dump of users Cards). Where is the exported file stored. I can't find it anywhere. 

Would it function with the command deck:import on the new NextCloud ?

**Comment by gabmert** [comment]
Hi @Majed610, when you use `occ deck:export` the info gets dumped to standard output, usually your terminal. If you want to have it in a file, you could use following command, which is [bash redirection]( https://www.gnu.org/software/bash/manual/bash.html#Redirections) `occ deck:export > ~/deck-export.json`. 

If you follow [above post](https://github.com/nextcloud/deck/issues/87#issuecomment-1005058965) you can see that the import function doesn't seem to be implemented yet.

**Comment by ne0YT** [comment]
so this (import) still just doesn't work? if so.. why are those issues closed? why are the docs suggesting otherwise? and if not.. how tf would I import?

**Comment by sunjam** [comment]
The remaining current issues were linked above, but here they are directly:
- https://github.com/nextcloud/deck/issues/3508
- https://github.com/nextcloud/deck/issues/3506

**Comment by ne0YT** [comment]
@sunjam still there is no functional import using the ci. those remaining issues are about a ui feature

**Comment by sunjam** [comment]
All I know is they are occ commands for import and export.  Listed above or in docs. https://deck.readthedocs.io/en/latest/User_documentation_en/#6-import-boards
